File: implementations/csharp/client/src/Specus.Client/PeerMesh/SecretFileWriter.cs
using System.Runtime.InteropServices;
using System.Text;

namespace Specus.Client.PeerMesh;

/// <summary>
/// Writes a secret to disk so no other local account can read it and no reader can observe a
/// half-written file.
///
/// <para>Writing straight to the destination has two failure modes. If the process dies mid-write
/// the client is left with a truncated private key and silently loses its peer identity on next
/// start. And the file exists, briefly, before anything narrows its permissions. Writing to a
/// temporary file that is locked down before the secret goes into it, then renaming it into place,
/// closes both: the permissions are never wrong even for an instant, and the rename is atomic, so a
/// reader sees either the old file or the complete new one.</para>
/// </summary>
internal static class SecretFileWriter
{
    private const UnixFileMode OwnerOnlyFile = UnixFileMode.UserRead | UnixFileMode.UserWrite;
    private const UnixFileMode OwnerOnlyDirectory =
        UnixFileMode.UserRead | UnixFileMode.UserWrite | UnixFileMode.UserExecute;

    /// <summary>Creates the directory if needed and restricts it to the current user.</summary>
    internal static void CreatePrivateDirectory(string directory)
    {
        if (Directory.Exists(directory))
        {
            Restrict(directory, OwnerOnlyDirectory);
            return;
        }
        if (OperatingSystem.IsWindows())
        {
            Directory.CreateDirectory(directory);
        }
        else
        {
            Directory.CreateDirectory(directory, OwnerOnlyDirectory);
        }
        Restrict(directory, OwnerOnlyDirectory);
    }

    /// <summary>Atomically writes <paramref name="content"/>, readable only by the current user.</summary>
    internal static void WriteSecret(string path, string content)
    {
        var directory = Path.GetDirectoryName(Path.GetFullPath(path))
                        ?? throw new IOException($"secret path has no parent directory: {path}");
        CreatePrivateDirectory(directory);

        // The temporary file lives in the destination directory so the replace stays on one volume.
        var temporary = Path.Combine(directory, $".secret-{Guid.NewGuid():N}.tmp");
        try
        {
            if (OperatingSystem.IsWindows())
            {
                File.WriteAllText(temporary, content, Encoding.UTF8);
            }
            else
            {
                // Create with the restrictive mode so the secret is never written into a file that
                // was, even momentarily, group or world readable.
                using (var stream = new FileStream(temporary, new FileStreamOptions
                       {
                           Mode = FileMode.CreateNew,
                           Access = FileAccess.Write,
                           UnixCreateMode = OwnerOnlyFile,
                       }))
                using (var writer = new StreamWriter(stream, new UTF8Encoding(false)))
                {
                    writer.Write(content);
                }
            }
            Restrict(temporary, OwnerOnlyFile);

            if (File.Exists(path))
            {
                File.Replace(temporary, path, destinationBackupFileName: null);
            }
            else
            {
                File.Move(temporary, path);
            }
            Restrict(path, OwnerOnlyFile);
        }
        finally
        {
            if (File.Exists(temporary))
            {
                try
                {
                    File.Delete(temporary);
                }
                catch (IOException)
                {
                    // Losing a temp file is not worth failing the write that already succeeded.
                }
            }
        }
    }

    private static void Restrict(string path, UnixFileMode mode)
    {
        if (OperatingSystem.IsWindows())
        {
            RestrictWindowsAcl(path);
            return;
        }
        try
        {
            File.SetUnixFileMode(path, mode);
        }
        catch (Exception error) when (error is PlatformNotSupportedException or UnauthorizedAccessException)
        {
            // A filesystem without mode bits (a mounted share) cannot be tightened further here.
        }
    }

    /// <summary>
    /// Replaces the inherited ACL with one granting the current user alone. Unix mode bits are
    /// ignored on Windows, so without this the file inherits whatever the profile directory grants,
    /// which on a shared or domain-joined machine can include other principals.
    /// </summary>
    private static void RestrictWindowsAcl(string path)
    {
        if (!OperatingSystem.IsWindows())
        {
            return;
        }
        try
        {
            var startInfo = new System.Diagnostics.ProcessStartInfo("icacls")
            {
                CreateNoWindow = true,
                UseShellExecute = false,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
            };
            startInfo.ArgumentList.Add(path);
            startInfo.ArgumentList.Add("/inheritance:r");
            startInfo.ArgumentList.Add("/grant:r");
            startInfo.ArgumentList.Add($"{Environment.UserName}:F");

            using var process = System.Diagnostics.Process.Start(startInfo);
            process?.WaitForExit(10_000);
        }
        catch (Exception error) when (error is System.ComponentModel.Win32Exception or InvalidOperationException)
        {
            // Failing to tighten the ACL must not stop the client from having a key at all; the
            // file is still created under the user profile.
        }
    }
}

File: implementations/java/client/src/main/java/com/theshuai/specusclient/peer/SecretFileWriter.java
package com.theshuai.specusclient.peer;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.FileSystems;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.nio.file.attribute.AclEntry;
import java.nio.file.attribute.AclEntryPermission;
import java.nio.file.attribute.AclEntryType;
import java.nio.file.attribute.AclFileAttributeView;
import java.nio.file.attribute.PosixFilePermission;
import java.nio.file.attribute.PosixFilePermissions;
import java.nio.file.attribute.UserPrincipal;
import java.util.EnumSet;
import java.util.List;
import java.util.Set;

/**
 * Writes a secret to disk so that no other local account can read it and no reader can ever observe
 * a half-written file.
 *
 * <p>Two separate problems are being solved. Writing directly to the destination leaves the file
 * world-readable for the window between creation and any later permission change, and leaves a
 * truncated key behind if the process dies mid-write — on next start the client would read a
 * corrupt private key and silently lose its peer identity. Writing to a temporary file that is
 * locked down <em>before</em> the secret goes into it, then renaming it into place, closes both:
 * the permissions are never wrong even briefly, and the rename is atomic, so a reader sees either
 * the old file or the complete new one.
 */
final class SecretFileWriter {
    private static final Set<PosixFilePermission> OWNER_ONLY_FILE =
            PosixFilePermissions.fromString("rw-------");
    private static final Set<PosixFilePermission> OWNER_ONLY_DIRECTORY =
            PosixFilePermissions.fromString("rwx------");

    private SecretFileWriter() {
    }

    /** Creates the directory if needed and restricts it to the current user. */
    static void createPrivateDirectory(Path directory) throws IOException {
        if (!Files.exists(directory)) {
            if (posixSupported()) {
                Files.createDirectories(directory,
                        PosixFilePermissions.asFileAttribute(OWNER_ONLY_DIRECTORY));
                return;
            }
            Files.createDirectories(directory);
            restrictWindowsAcl(directory);
            return;
        }
        restrictExisting(directory, OWNER_ONLY_DIRECTORY);
    }

    /** Atomically writes {@code content}, readable only by the current user. */
    static void writeSecret(Path target, String content) throws IOException {
        Path directory = target.toAbsolutePath().getParent();
        if (directory == null) {
            throw new IOException("secret path has no parent directory: " + target);
        }
        createPrivateDirectory(directory);

        // The temporary file lives in the destination directory so the rename stays on one
        // filesystem, which is what makes it atomic.
        Path temporary = Files.createTempFile(directory, ".secret", ".tmp");
        try {
            restrictExisting(temporary, OWNER_ONLY_FILE);
            Files.writeString(temporary, content, StandardCharsets.UTF_8);
            try {
                Files.move(temporary, target,
                        StandardCopyOption.REPLACE_EXISTING, StandardCopyOption.ATOMIC_MOVE);
            } catch (IOException atomicUnsupported) {
                // Some filesystems refuse ATOMIC_MOVE. A plain replace is still better than
                // writing the secret into the destination directly.
                Files.move(temporary, target, StandardCopyOption.REPLACE_EXISTING);
            }
            restrictExisting(target, OWNER_ONLY_FILE);
        } finally {
            Files.deleteIfExists(temporary);
        }
    }

    private static void restrictExisting(Path path, Set<PosixFilePermission> permissions)
            throws IOException {
        if (posixSupported()) {
            Files.setPosixFilePermissions(path, permissions);
            return;
        }
        restrictWindowsAcl(path);
    }

    private static boolean posixSupported() {
        return FileSystems.getDefault().supportedFileAttributeViews().contains("posix");
    }

    /**
     * Replaces the inherited ACL with a single entry for the owner. Without this the file inherits
     * whatever the parent directory grants, which on a shared machine can include other users.
     */
    private static void restrictWindowsAcl(Path path) throws IOException {
        AclFileAttributeView view = Files.getFileAttributeView(path, AclFileAttributeView.class);
        if (view == null) {
            return;
        }
        UserPrincipal owner = Files.getOwner(path);
        if (owner == null) {
            return;
        }
        AclEntry entry = AclEntry.newBuilder()
                .setType(AclEntryType.ALLOW)
                .setPrincipal(owner)
                .setPermissions(EnumSet.allOf(AclEntryPermission.class))
                .build();
        view.setAcl(List.of(entry));
    }
}
